src/librustc/mir/visit.rs

```diff
@@ -675,20 +675,12 @@ macro_rules! make_mir_visitor {
                             place: & $($mutability)? Place<'tcx>,
                             context: PlaceContext,
                             location: Location) {
-                match place {
-                    Place::Base(place_base) => {
-                        self.visit_place_base(place_base, context, location);
-                    }
-                    Place::Projection(proj) => {
-                        let context = if context.is_mutating_use() {
-                            PlaceContext::MutatingUse(MutatingUseContext::Projection)
-                        } else {
-                            PlaceContext::NonMutatingUse(NonMutatingUseContext::Projection)
-                        };
+                let Place { base, projection } = place;
 
-                        self.visit_place(& $($mutability)? proj.base, context, location);
-                        self.visit_projection(proj, location);
-                    }
+                self.visit_place_base(base, context, location);
+
+                if let Some(proj) = projection {
+                    self.visit_projection(proj, location);
```

Comment:
the old code recursively visited all projections. I think you should use iterate here in order to visit all of them, too.

This may in fact be what causes the current ICE, some types aren't getting visited.

Reply:
Right, wrong rebase. Done.